### PR TITLE
Add NeRF Synthetic dataloader

### PR DESCRIPTION
**Pull request by G-1nOnly**
Hello,

Thanks for such a great work!

I have added the dataloader for NeRF Synthetic dataset as suggested in this [issue](https://github.com/theialab/radfoam/issues/14). The dataloader works from my side and I also load the image mask of the dataset as input.

By using this dataloader, radfoam can be used for training scenes in NeRF Synthetic dataset, however, the results are not okay on the first place, adding mask loss can achieve better quality. The results would be like the following figure.

rgb_000_psnr_33 274

### CODE REVIEW

**Comment by Shrisudhan** [comment]
Hi, can you also also commit the config file and the mask loss so that I can test it once and merge it into the main branch.

**Comment by G-1nOnly** [comment]
Thanks for the instant reply!

Of course! I've just commit the train file with mask loss, which I set the weight as 0.01 for now and can be move to pipeline_args later. I've also committed the config file though I haven't changed any parameters, which I think some needs to be tuned.

**Comment by Shrisudhan** [comment]
Got it. I will test it today and get back to you.

**Comment by Shrisudhan** [comment]
Hi, @G-1nOnly, I added the dataloader for blender scenes `77f5cc8`. I made some minor changes to be compatible with latest commits in dataloader. I tried the entropy loss you proposed to use for blender scenes. I couldn't get it to working well. In `77f5cc8` I added support so we can apply L2 loss between ray opacity and alpha mask from the blender dataset. This is still not working as good as I want. It leads to some boundary artifacts as you can see here: 
rgb_070_psnr_26 346

I will try to work on that a bit more when I get some time. But in case you want to explore more, this is the config file I used for this run:
```
# Model Parameters
sh_degree: 3
init_points: 8_192
final_points: 524_288
activation_scale: 1
device: cuda

# Pipeline Parameters
iterations: 10_000
densify_from: 1_000
densify_until: 6_000
densify_factor: 1.15
white_background: true
quantile_weight: 0
viewer: false                          # Flag to use viewer
debug: false                           # Flag to not use tensorboard

# Optimization Parameters
points_lr_init: 2e-4
points_lr_final: 5e-6
density_lr_init: 1e-1
density_lr_final: 1e-2
attributes_lr_init: 5e-3
attributes_lr_final: 5e-4
sh_factor: 0.1
freeze_points: 8_000                  # Points are frozen after this cycle

# Dataset Parameters
dataset: "blender"
data_path: "data/blender"
scene: "lego"
patch_based: false
downsample: [2, 1]                     # Image downsample factors
downsample_iterations: [0, 1_000]
```

**Comment by G-1nOnly** [comment]
Thank you for the reply and the latest commit. Indeed, after I implemented the entropy loss, there are still some floating artifacts. I'll try to work on it a bit more and let you know if I make any progress.

**Comment by G-1nOnly** [comment]
I use the dataloader from my commit and use the same config you provided and with the same loss you used, I got the following results.
rgb_070_psnr_32 233

It seems that the result is okay, but there are some floating artifacts shown as follows. These artifacts only appear in some views,  and I would try to improve the results.
rgb_082_psnr_19 421

**Comment by Shrisudhan** [comment]
Hey, thanks for testing it out. I guess there is a bug in my dataloader. Let me try to fix that. Meanwhile, can you try changing the [standard deviation of the normal initialization](https://github.com/theialab/radfoam/blob/main/radfoam_model/scene.py#L63) to 1? Also, bringing back the quantile loss will help remove floaters(you can change this in config, `quantile_weight: 1e-4`).

**Comment by G-1nOnly** [comment]
Thanks for the advice. I've tried setting the standard deviation of the normal initialization to 1 and adding the quantile loss. I found that even without the quantile loss, the floating artifacts can be eliminated. I think it's probably due to a rather compact initialization after setting it to 1. The result is shown as follows. 

rgb_083_psnr_27 500

Though there exists some blurry parts, the result is much better. By adding the quantile loss, I get slightly lower average PSNR but better quality for these views with blurry parts. The result is shown below.

rgb_083_psnr_27 651

 I would try to further improve the results.

**Comment by Shrisudhan** [comment]
Hey, you are right. I had a bug in my alpha blending equation for the data loader. I fixed it in the latest commit. I am able to achieve 33.5 dB PSNR with this configuration. The results look visually fine(tested in the viewer as well). I guess need to do some hyper-parameter tuning for the PSNR.
```
# Model Parameters
sh_degree: 3
init_points: 8_192
final_points: 524_288
activation_scale: 1
device: cuda

# Pipeline Parameters
iterations: 20_000
densify_from: 2_000
densify_until: 11_000
densify_factor: 1.15
white_background: true
quantile_weight: 0
viewer: false                          # Flag to use viewer
debug: false                           # Flag to not use tensorboard

# Optimization Parameters
points_lr_init: 2e-4
points_lr_final: 5e-6
density_lr_init: 1e-1
density_lr_final: 1e-2
attributes_lr_init: 5e-3
attributes_lr_final: 5e-4
sh_factor: 0.02
freeze_points: 18_000                  # Points are frozen after this cycle

# Dataset Parameters
dataset: "blender"
data_path: "data/blender"
scene: "lego"
patch_based: false
downsample: [2, 1]                     # Image downsample factors
downsample_iterations: [0, 2_000]
```

**Comment by G-1nOnly** [comment]
Thanks for the instant reply, I can also achieve the same result with the same setting. I guess the dataloader works okay for now, and if you feel it's okay then we can close this pull request.

**Comment by Shrisudhan** [comment]
Okay, sure. If you are able to make the PSNR go higher, we can create a new PR and add the config files as well. Let me know if you need any help with that.
